src/utils/s3/s3_bulk_renaming.py

```diff
@@ -0,0 +1,130 @@
+import argparse
+import re
+import sys
+from typing import Optional
+
+import boto3  # type: ignore
+from botocore.exceptions import NoCredentialsError  # type: ignore
+
+
+def list_objects(bucket_name: str, prefix: Optional[str] = None) -> list[str]:
+    """
+    Lists all objects in the specified S3 bucket matching the given prefix.
+    Returns a list of object keys.
+    """
+    s3 = boto3.client("s3")
+    keys: list[str] = []
+
+    paginator = s3.get_paginator("list_objects_v2")
+    page_iterator = paginator.paginate(Bucket=bucket_name, Prefix=prefix)
+
+    try:
+        for page in page_iterator:
+            if "Contents" in page:
+                keys.extend(obj["Key"] for obj in page["Contents"])
+    except NoCredentialsError:
```

Comment:
I think I would suggest letting this exception bubble out of `list_objects` and put your try-except block with `sys.exit` in `main()`.

One reason to break up these scripts into functions instead of just one imperative script that executes start-to-finish is that it makes it possible to use the module as a library or to execute it directly. This is a pattern we see in the Python standard library with modules like `http` where you can import it to build an HTTP server or run it on the commandline to start up a simple HTTP server.

So if we think about using this module as a library, catching the `NoCredentialsError` and calling `sys.exit` could really interfere with someone’s ability to use this function in another program.

Reply:
That's a great point. Thanks for the suggestion!